irods/connection.py

```diff
@@ -149,7 +149,9 @@ def recv(self, into_buffer = None
             except TypeError:
                 err_msg = None
             if nominal_code(msg.int_info) not in acceptable_codes:
-                raise get_exception_by_code(msg.int_info, err_msg)
+                exc = get_exception_by_code(msg.int_info, err_msg)
+                exc.server_msg = msg
+                raise exc
```

Comment:
Is `msg` guaranteed to be valid here?

Reply:
Guaranteed to exist, but I don't know about validity yet.  It will be whatever message the server sent back, translated into Pythonic form.

Reply:
Guaranteed to exist is good.

Will this affect the outcome of `.server_msg.get_json_encoded_struct()`?

Reply:
`msg` is not guaranteed to be a valid target for the method `get_json_encoded_struct`
At times the returned JSON message is embedded and base64 encoded in a BytesBuf_PI (such as with the Atomic Metadata API)  and at other times it comes via a STR_PI (for example, with the replica truncate API).  

So, `get_json_encoded_struct( )` seems not to be a dependable way in general to arrive at the JSON formatted response for all base classes of `iRODSException`.  When it's a STR_PI that resides in the `server_msg`, you need `server_msg.get_main_message(STR_PI).myStr` to arrive at the JSON contents. 

Here's an example of the msg component of the server_msg object for the atomic metadata API
```
b'<BinBytesBuf_PI>\n<buflen>144</buflen>\n<buf>eyJlcnJvcl9tZXNzYWdlIjoiRW50aXR5IGRvZXMgbm90IGV4aXN0IFtlbnRpdHlfbmFtZT0vdGVtcFpvbmUvaG9tZS9yb2RzL3Rlc3RfZGlyL2F0b21pY19tZXRhX2lzc3VlXzYwNl0iLCJvcGVyYXRpb24iOm51bGwsIm9wZXJhdGlvbl9pbmRleCI6MH0A</buf>\n</BinBytesBuf_PI>\n'
```
And here's an example of the corresponding member for replica truncate:
```
b'<STR_PI>\n<myStr>{&quot;message&quot;:&quot;rs_replica_truncate: Hierarchy descending from specified resource name [aaa] does not have a replica of [/tempZone/home/rods/abc.dat] or the replica is inaccessible at this time.&quot;,&quot;replica_number&quot;:null,&quot;resource_hierarchy&quot;:null}</myStr>\n</STR_PI>\n'
```
Perhaps a better method can be written, one that improves on` get_json_encoded_struct` and works for the general case.  The only alternative is, of course, for the application writer to be well versed on the details of parsing each possible subtype of an `iRODSException`.

Reply:
I believe I've done this, so I'll resolve for now.  As long as an api places its output in a `char**`, `BinBytesBuf_PI**` or `BytesBuf_PI**` and that output is json-formatted, `get_json_encoded_struct` will work.